program/src/processor/inactivate_stake.rs

```diff
@@ -13,8 +21,91 @@ use crate::instruction::accounts::{Context, InactivateStakeAccounts};
 /// 0. `[w]` Stake config account
 /// 1. `[w]` Validator stake account
 pub fn process_inactivate_stake(
-    _program_id: &Pubkey,
-    _ctx: Context<InactivateStakeAccounts>,
+    program_id: &Pubkey,
+    ctx: Context<InactivateStakeAccounts>,
 ) -> ProgramResult {
+    // Account valuidation.
+
+    // config
+    // - owner must be the stake program
+    // - must be initialized
+
+    require!(
+        ctx.accounts.config.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "config"
+    );
+
+    let data = &mut ctx.accounts.config.try_borrow_mut_data()?;
+    let config = bytemuck::try_from_bytes_mut::<Config>(data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        config.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "config",
+    );
+
+    // stake
+    // - owner must be the stake program
+    // - must be initialized
+    // - must have the correct derivation
+
+    require!(
+        ctx.accounts.stake.owner == program_id,
+        ProgramError::InvalidAccountOwner,
+        "stake"
+    );
+
+    let data = &mut ctx.accounts.stake.try_borrow_mut_data()?;
+    let stake = bytemuck::try_from_bytes_mut::<Stake>(data)
+        .map_err(|_error| ProgramError::InvalidAccountData)?;
+
+    require!(
+        stake.is_initialized(),
+        ProgramError::UninitializedAccount,
+        "stake",
+    );
+
+    // validates that the stake account corresponds to the received
+    // config account
+    let (derivation, _) = find_stake_pda(&stake.validator, ctx.accounts.config.key, program_id);
+
+    require!(
+        ctx.accounts.stake.key == &derivation,
+        ProgramError::InvalidSeeds,
+        "stake",
+    );
+
+    // Inactivates the stake if elegible.
+    //
+    // Note: we do not fail if there are no tokens to inactivate
+
+    if let Some(timestamp) = stake.deactivation_timestamp {
+        let current = Clock::get()?
+            .unix_timestamp
+            .saturating_sub(config.cooldown_time_seconds) as u64;
+
+        // TODO: should we fail or just do nothing?
+        require!(
+            current >= timestamp.get(),
+            StakeError::ActiveDeactivationCooldown
+        );
+
+        msg!("Deactivating {} token(s)", stake.deactivating_amount);
+
+        config.token_amount_delegated = config
+            .token_amount_delegated
+            .saturating_sub(stake.deactivating_amount);
```

Comment:
That is a good question. My interpretation was to decrease the total once the tokens are "moved" to "inactive" – although they only really move on the `WithdrawInactiveStake` as you pointed out. So the `inactive` tokens are not eligible for rewards independently from whether they have been withdrawn or not.

If we don't decrease the `token_amount_delegated`, then we should not decrease the `amount` either.

Reply:
Had another look at this and we could do the following:
- `config.token_amount_delegated`: it is the total number of tokens (`amount` + `inactive_amount`), only decreases on `WithdrawInactiveStake`.
- `stake.amount`: it is only the "active" amount, it decreases when token are "moved" from deactivated to inactive.

Reply:
I think this is correct yeah, since `WithdrawInactiveStake` is documented as saying it's moving tokens "out of the vault", which is represented by that `token_amount_delegated`.